erdos/ray/ray_operator.py

```diff
@@ -52,15 +52,46 @@ def on_completion_msg(self, msg):
         """Invokes corresponding callback for stream stream_name."""
         self._op.log_event(time.time(), msg.timestamp,
                            'receive watermark {}'.format(msg.stream_name))
-        for cb in self._completion_callbacks.get(msg.stream_uid, []):
-            cb(msg)
+
+        # Ensure that the watermark is monotonically increasing.
+        high_watermark = self._op._stream_to_high_watermark[msg.stream_name]
+        if not high_watermark:
+            # The first watermark, just set the dictionary with the value.
+            self._op._stream_to_high_watermark[
+                msg.stream_name] = msg.timestamp
+        else:
+            if high_watermark >= msg.timestamp:
+                raise Exception(
+                    "The watermark received in the msg {} is not "
+                    "higher than the watermark previously received "
+                    "on the same stream: {}".format(msg, high_watermark))
+            else:
+                self._op._stream_to_high_watermark[msg.stream_name] = \
+                        msg.timestamp
+
+        # Now check if all other streams have a higher or equal watermark.
+        # If yes, flow this watermark. If not, return from this function
+        # Also, maintain the lowest watermark observed.
+        low_watermark = msg.timestamp
+        for stream, watermark in self._op._stream_to_high_watermark.items():
+            if stream != msg.stream_name:
+                if not watermark or watermark < msg.timestamp:
+                    return
+                if low_watermark > watermark:
```

Comment:
I think there's a case here when you'll end up duplicating watermarks. Input stream i1 received watermark t, i2 received watermark t (and you issued downstream watermark t). Now if i2 receives t +1, you'll send again downstream watermark t.

IMHO, the code would be easier to read if you kept a global low watermark, do a loop over the watermarks to get the min watermark on the inputs. If this min is greater than the global, update and send, otherwise return.

Reply:
I don't think the situation you pointed out is possible. The `if not watermark or watermark < msg.timestamp` protects against that.
So, when i1 receives a watermark t, and i2 has no initial watermark, the `if not watermark` clause will be invoked and the method will return.
When i2 receives a watermark t, i1 will have the watermark t and the `not watermark or watermark < msg.timestamp` will evaluate to `False`, and thus the timestamp t will flow.
When i2 receives a watermark t+1, `watermark < msg.timestamp` will evaluate `True` for i1 and we will not flow a watermark downstream.